pkg/catalogv2/git/git.go

```diff
@@ -1,397 +1,480 @@
-//   - Otherwise, it sets the RefSpec to fetch all branches.
-//
-// fetching the last commit of one branch is faster than fetching from all branches.
-func (r *Repository) updateRefSpec(branch string) {
-	var newRefSpec string
-
-	if branch != "" {
-		newRefSpec = fmt.Sprintf("+refs/heads/%s:refs/remotes/origin/%s", branch, branch)
-	} else {
-		newRefSpec = "+refs/heads/*:refs/remotes/origin/*"
+func (g *git) clone(branch string) error {
+	gitDir := filepath.Join(g.Directory, ".git")
+	if dir, err := os.Stat(gitDir); err == nil && dir.IsDir() {
+		return nil
 	}
 
-	if len(r.fetchOpts.RefSpecs) > 0 {
-		r.fetchOpts.RefSpecs[0] = config.RefSpec(newRefSpec)
-	} else {
-		r.fetchOpts.RefSpecs = []config.RefSpec{config.RefSpec(newRefSpec)}
+	if err := os.RemoveAll(g.Directory); err != nil {
+		return fmt.Errorf("failed to remove directory %s: %v", g.Directory, err)
 	}
-}
 
-// fetch fetches updates from the remote repository for a specific branch.
-// If the fetch operation is already up-to-date, this is not treated as an error.
-// Any other error that occurs during fetch is returned.
-func (r *Repository) fetch(branch string) error {
-	r.updateRefSpec(branch)
-	fetchOptions := r.fetchOpts
+	return g.Clone(branch)
+}
 
-	err := r.repoGogit.Fetch(fetchOptions)
-	if err != nil && err != gogit.NoErrAlreadyUpToDate && err != transport.ErrEmptyUploadPackRequest {
-		return fmt.Errorf("fetch failure: %w", err)
+func (g *git) fetchAndReset(rev string) error {
+	if err := g.git("-C", g.Directory, "fetch", "origin", "--", rev); err != nil {
+		return err
 	}
+	return g.reset("FETCH_HEAD")
+}
 
-	return nil
+func (g *git) reset(rev string) error {
+	return g.git("-C", g.Directory, "reset", "--hard", rev)
+}
+
+func (g *git) currentCommit() (string, error) {
+	return g.gitOutput("-C", g.Directory, "rev-parse", "HEAD")
 }
 
-// hardReset performs a hard reset of the git repository to a specific commit.
-func (r *Repository) hardReset(reference string) error {
-	var err error
-	resetOpts := r.resetOpts
+func (g *git) gitCmd(output io.Writer, args ...string) error {
+	kv := fmt.Sprintf("credential.helper=%s", `/bin/sh -c 'echo "password=$GIT_PASSWORD"'`)
+	cmd := exec.Command("git", append([]string{"-c", kv}, args...)...)
+	cmd.Env = append(os.Environ(), fmt.Sprintf("GIT_PASSWORD=%s", g.password))
+	stderrBuf := &bytes.Buffer{}
+	cmd.Stderr = stderrBuf
+	cmd.Stdout = output
 
-	switch {
-	case isLocalBranch(reference):
-		branchRef, err := r.repoGogit.Reference(plumbing.ReferenceName(reference), true)
+	if g.agent != nil {
+		c, err := g.injectAgent(cmd)
 		if err != nil {
-			return fmt.Errorf("hardReset failure, branch does not exist locally: %w", err)
+			return err
 		}
-		resetOpts.Commit = branchRef.Hash()
-	case reference == plumbing.HEAD.String():
-		commitHash, err := r.getCurrentCommit()
+		defer c.Close()
+	}
+
+	if len(g.knownHosts) != 0 {
+		f, err := ioutil.TempFile("", "known_hosts")
```

Comment:
ioutil.TempDir is deprecated: As of Go 1.17, this function simply calls os.MkdirTemp.

If the CI does not break because of this it is ok.